### PR TITLE
Bug 1102834 - Save UserDictionary TST blob in KeyboardSettingsApp

### PR DESCRIPTION
**Pull request by mnjul**
- An empty word list saves into indexedDB as "undefined" in TST blob
- Add WordListConverter which is the JavaScript version of xml2dict.py
  -- More information: https://github.com/mnjul/gaia-kb-dict-tools/tree/release/gaiafication
- Change UserDictionary._saveDict() unit test script such that testing queued saves is more realistic and more streamlined.
- Implement PromiseStorage.setItems()

### CODE REVIEW

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b68912ec653f)

File: apps/keyboard/js/settings/tst_converter.js [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+  return offset;
+};
+
+// JSConv:
+// In the JS version, since we're not directly writing to a file,
+// 'output' is a JS array. We convert to UInt8Array when we
+// finishes pushing to 'output'. This is because UInt8Array's length
+// has to be decided at instantiation.
+// XXX: See if we can pre-determine the length when instantiating
+//      the buffer.
```
Comment (mnjul) [inline apps/keyboard/js/settings/tst_converter.js]: After I've now studied the blob format and generation code more thoroughly, I believe this pre-determination is quite possible. I'll do that in a follow-up bug.

File: apps/keyboard/js/settings/user_dictionary.js [hunk 2]
```diff
@@ -52,14 +52,22 @@ UserDictionary.prototype.getList = function() {
...
+UserDictionary.prototype._saveDict = function() {
+  var wordList = Array.from(this._wordSet);
+
+  var dictBlob = new TSTConverter(wordList).toBlob();
+
   var p = this._saveQueue.then(
-    () => this._dbStore.setItem('wordlist', Array.from(this._wordSet))
+    () => this._dbStore.setItem('wordlist', wordList)
+  ).then(
+    () => this._dbStore.setItem('dictblob', dictBlob)
```
Comment (timdream) [inline apps/keyboard/js/settings/user_dictionary.js]: How about implement `Promise#setItems` and use that here? I am sorry that I didn't do that back then ... I thought I did.

Comment (mnjul) [inline apps/keyboard/js/settings/user_dictionary.js]: Yep, sure, I'll implement the function.

File: apps/keyboard/test/unit/settings/user_dictionary_test.js [hunk 3]
```diff
@@ -177,57 +184,120 @@ suite('UserDictionary', function() {
     assert.deepEqual(Array.from(model._wordSet), []);
   });
 
-  test('saveList', function() {
-    // we test successive three calls to saveList and make sure they're properly
+  test('saveDict', function() {
```
Comment (mnjul) [inline apps/keyboard/test/unit/settings/user_dictionary_test.js]: Let's investigate if we can simply the test codes by using a real Promise.

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=11eba7b18769)

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c41616ab3be8)

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=a5f6be472b18)

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=17737940429c)

File: apps/keyboard/js/settings/word_list_converter.js [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+
+
+
+// JSConv:
+// In the JS version, since we're not directly writing to a file,
+// 'output' is a JS array. We convert to UInt8Array when we
+// finishes pushing to 'output'. This is because UInt8Array's length
+// has to be decided at instantiation.
+// XXX: See if we can pre-determine the length when instantiating
+//      the buffer.
```
Comment (mnjul) [inline apps/keyboard/js/settings/word_list_converter.js]: After I've now studied the blob format and generation code more thoroughly, I believe this pre-determination is quite possible. I'll do that in a follow-up bug.

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b86339147516)

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=d25ab65e3d76)

**Comment by autolander** [comment]
The pull request could not be applied to the integration branch. Please try again after current integration is complete.

File: apps/keyboard/js/settings/word_list_converter.js (hunk 4 above)
Comment (mnjul) [inline apps/keyboard/js/settings/word_list_converter.js]: And that is https://bugzilla.mozilla.org/show_bug.cgi?id=1130317

**Pull request merged**